blockchain/rolling_merkle.go

```diff
@@ -0,0 +1,149 @@
+package blockchain
+
+import (
+	"crypto/sha256"
+	"math/bits"
+
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+)
+
+// parentHash returns the hash of the left and right hashes passed in.
+func parentHash(l, r chainhash.Hash) chainhash.Hash {
+	h := sha256.New()
+	h.Write(l[:])
+	h.Write(r[:])
+
+	buf := make([]byte, 0, 32)
+	first := h.Sum(buf)
+	h.Reset()
+	h.Write(first)
+	return *((*chainhash.Hash)(h.Sum(buf)))
+}
+
+// rollingMerkleTreeStore calculates the merkle root by only allocating O(logN)
+// memory where N is the total amount of leaves being included in the tree.
+type rollingMerkleTreeStore struct {
+	// roots are where the temporary merkle roots get stored while the
+	// merkle root is being calculated.
+	roots []chainhash.Hash
+
+	// numLeaves is the total leaves the store has processed.  numLeaves
+	// is required for the root calculation algorithm to work.
```

Comment:
Right but you never clear the values in the roots slice afaict so even though there is data there, any root value that corresponds to an *internal* zero is conceptually garbage with respect to the algorithm here. My comment about truncating the trailing zeroes is indeed wrong though because 1000 still has 3 garbage values in the roots slice. I went to verify go's slice popping behavior, and changing the length the way you do in the loop doesn't zero the memory of what was there before. Maybe you think that including this detail in the comments is unnecessary, but I am fairly confident that the claims I make in this comment are correct.

Reply:
FYI this thread is nonblocking

Reply:
Won't be addressing this in this PR